### PR TITLE
feat: the overall optimization and architecture design of the makefile are made

### LINKED ISSUE #305: feature: Provide a better development experience
## Proposal

It's great to see such a promising, future-oriented, and active project, especially with K8sGPT's superpowers like ChatGPT.

As a Kubernetes enthusiast, I was deeply attracted to K8sGPT, so I pulled the code to my local machine to dive deeper into it.

However, I encountered some issues:

- The Makefile doesn't provide compilation and testing capabilities, which is currently very simple:
    
    ```
    ❯ cat Makefile
    .PHONY: docker-build
    IMG ?= ghcr.io/k8sgpt-ai/k8sgpt:latest
    
    deploy:
    ifndef SECRET
            $(error SECRET environment variable is not set)
    endif
            kubectl create ns k8sgpt || true
            kubectl create secret generic ai-backend-secret --from-literal=secret-key=$(SECRET) --namespace=k8sgpt || true
            kubectl apply -f container/manifests
    undeploy:
            kubectl delete secret ai-backend-secret --namespace=k8sgpt
            kubectl delete -f container/manifests
            kubectl delete ns k8sgpt
    docker-build:
            docker buildx build --build-arg=VERSION="$$(git describe --tags --abbrev=0)" --build-arg=COMMIT="$$(git rev-parse --short HEAD)" --build-arg DATE="$$(date +%FT%TZ)" --platform="linux/amd64,linux/arm64" -t ${IMG} -f container/Dockerfile . --push
    
    ```
    
- The handling of the binary after compilation is not very friendly, even though the [[contributing document](https://github.com/k8sgpt-ai/k8sgpt/blob/main/CONTRIBUTING.md)](https://github.com/k8sgpt-ai/k8sgpt/blob/main/CONTRIBUTING.md) tells me: Building the binary is as simple as running `go build .` in the root of the repository.

## Suggestions

As my first step in contributing to K8sGPT, I would like to hear everyone's suggestions for any good features and proposals.

### PR DESCRIPTION
**Pull request by cubxxw**
## feat: the overall optimization and architecture design of the makefile are made 

Closes #305 

## 📑 Description

Read the proposal #305 for the Makefile design

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [x] I have updated the documentation as required
- [x] All the tests have passed

## ℹ Additional Information
image

This will give you more information at build time:
```bash
❯ make build
go version go1.20 linux/amd64
===========> Building binary /root/workspaces/k8sgpt/bin/k8sgpt *[Git Info]: v0.2.4-13b2a1314f59d2a15a159a622c07596af22ffa31
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,16 +1,147 @@
...
+
+## vet: Run go vet against code.
+.PHONY: vet
+vet:
+	@$(GO) vet ./...
+
+## lint: Run go lint against code.
+.PHONY: lint
+lint:
+	@golangci-lint run -v ./...
```
Comment (matthisholleville) [inline Makefile:96]: Shouldn't verify `golangci-lint` is installed ?

```bash
## lint: Run go lint against code.
.PHONY: lint
lint:
	@if [ -x "$$(command -v golangci-lint)" ]; then \
		golangci-lint run -v ./... ; \
	else \
		echo "golangci-lint is not installed. Please install it using 'cd ../ && go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.52.2'" ; \
	fi
```

**Review by matthisholleville — COMMENTED** [review]
Very good job @cubxxw ! Left 1 comment, can't wait to hear what you think

**Comment by matthisholleville** [comment]
Hey, I noticed that you ticked the boxes:

- [x] My code requires changes to the documentation
- [x] I have updated the documentation as required

But I'm not seeing any changes in the documentation. Is everything cool or am I missing something?

File: Makefile (hunk 1 above)
Comment (cubxxw) [reply Makefile:96]: Yes, I think the installation should be verified and even judged, but I think one makefile is now too much. I would like to extend the script directory and divide it into makefile subdirectories

**Comment by cubxxw** [comment]
> Hey, I noticed that you ticked the boxes:
> 
> - [x] My code requires changes to the documentation
> - [x] I have updated the documentation as required
> 
> But I'm not seeing any changes in the documentation. Is everything cool or am I missing something?

Forgive me, I expect to find an opportunity to update later

I would like to know how to deal with the copyright, which will affect my next work

**Review by AlexsJones — APPROVED** [review]
This is a a super duper contribution, thank you!

**Pull request merged**